src/Surfnet/StepupGateway/SecondFactorOnlyBundle/Saml/ResponseFactory.php

```diff
@@ -0,0 +1,171 @@
+        $response->setDestination($targetServiceProvider->getAssertionConsumerUrl());
+        $response->setInResponseTo($this->proxyStateHandler->getRequestId());
+
+        return $response;
+    }
+
+    /**
+     * @param string $nameId
+     * @param ServiceProvider $targetServiceProvider
+     * @param string $authnContextClassRef
+     * @return SAML2_Assertion
+     */
+    private function createNewAssertion(
+        $nameId,
+        ServiceProvider $targetServiceProvider,
+        $authnContextClassRef
+    ) {
+        $newAssertion = new SAML2_Assertion();
+        $newAssertion->setNotBefore($this->currentTime->getTimestamp());
+        $newAssertion->setNotOnOrAfter($this->getTimestamp('PT5M'));
+        $newAssertion->setIssuer($this->hostedIdentityProvider->getEntityId());
+        $newAssertion->setIssueInstant($this->getTimestamp());
+        $this->assertionSigningService->signAssertion($newAssertion);
+        $this->addSubjectConfirmationFor($newAssertion, $targetServiceProvider);
+        $newAssertion->setNameId([
+            'Format' => \SAML2_Const::NAMEID_UNSPECIFIED,
+            'Value' => $nameId,
+        ]);
+        $newAssertion->setValidAudiences([$this->proxyStateHandler->getRequestServiceProvider()]);
+        $this->addAuthenticationStatementTo($newAssertion, $authnContextClassRef);
+
+        return $newAssertion;
+    }
+
+    /**
+     * @param SAML2_Assertion $newAssertion
+     * @param ServiceProvider $targetServiceProvider
+     */
+    private function addSubjectConfirmationFor(SAML2_Assertion $newAssertion, ServiceProvider $targetServiceProvider)
+    {
+        $confirmation         = new \SAML2_XML_saml_SubjectConfirmation();
+        $confirmation->Method = \SAML2_Const::CM_BEARER;
+
+        $confirmationData                      = new \SAML2_XML_saml_SubjectConfirmationData();
+        $confirmationData->InResponseTo        = $this->proxyStateHandler->getRequestId();
+        $confirmationData->Recipient           = $targetServiceProvider->getAssertionConsumerUrl();
+        $confirmationData->NotOnOrAfter        = $this->getTimestamp('PT8H');
+
+        $confirmation->SubjectConfirmationData = $confirmationData;
+
+        $newAssertion->setSubjectConfirmation([$confirmation]);
+    }
+
+    /**
+     * @param SAML2_Assertion $assertion
+     * @param SAML2_Assertion $assertion
+     */
+    private function addAuthenticationStatementTo(SAML2_Assertion $assertion, $authnContextClassRef)
+    {
+        $assertion->setAuthnInstant($this->getTimestamp());
+        $assertion->setAuthnContextClassRef($authnContextClassRef);
+        $assertion->setAuthenticatingAuthority([$this->hostedIdentityProvider->getEntityId()]);
+    }
+
+    /**
+     * @param string $interval a \DateInterval compatible interval to skew the time with
+     * @return int
+     */
+    private function getTimestamp($interval = null)
+    {
+        $time = clone $this->currentTime;
```

Comment:
I'm keeping the BFG in the closet for now as I'm apparently not familiar enough with the coding conventions in SSA.
What do you suggest I change here?